### PR TITLE
docs: extend examples suite

### LINKED ISSUE #274: Extend examples suite
MIssing examples:
- [ ] Multi model training.
- [ ] Custom logger example.
- [ ] Custom splitter in Dataset.split()
- [ ] Custom negative sampler.
- [ ] Custom model development and training with comparison to existing model.

Rename examples:
- [ ] `external_dataset` -> `load_dataset_from_external_source`

Reorg examples:
- hyperlink prediction (models)
- dataset (load, splitter)
- enrichment
- sampling
- ...

### PR DESCRIPTION
**Pull request by ddevin96**
### All Submissions

- [x] Have you followed the guidelines in our [Contributing](../../CONTRIBUTING.md) document?
- [x] Have you checked to ensure there are no other open [Pull Requests](../../../pulls) for the same changes?
- [x] Have you made sure you have rebased your branch to the latest commit on the target branch?

### Description

- Reorganise examples folder 
- Add examples for custom logger, splitter and negative sampler 
- Update script for running example 

Close #274 

### Checklist

- [x] Does your submission pass all tests? (use `make test`)
- [x] Have you written tests to cover all your changes? If not, provide a reason.
- [x] Have you lint your code locally before submission? (use `make format`)
- [x] Have you type checked your code locally before submission? (use `make typecheck`)

### CODE REVIEW

File: examples/train/multi_model_training.py
Comment (tizianocitro) [inline examples/train/multi_model_training.py:1]: I'd use `NHP` and `VilLain` as models for the example.

File: examples/splitting/custom_splitter.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    DefaultHDataSplitter,
+)
+from hyperbench.utils import (
+    NodeSpaceSetting,
+    is_transductive_setting,
+)
+import random
+
+
+class CustomSplitter(DefaultDatasetSplitter):
```
Comment (tizianocitro) [inline examples/splitting/custom_splitter.py]: Extend like:

```
CustomSplitter(Splitter["Dataset", tuple[list["Dataset"], list[float]]]):
```

And provide very simple `split` like splitting in half.

File: examples/train/train_with_custom_logger.py
Comment (tizianocitro) [inline examples/train/train_with_custom_logger.py:1]: Looks as an actual feature. I'd simply this logger, though.

File: examples/train/train_with_custom_logger.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+    with MultiModelTrainer(
+        model_configs=configs,
+        max_epochs=100,
+        accelerator="auto",
+        log_every_n_steps=10,
+        enable_checkpointing=False,
+        auto_start_tensorboard=True,
+        auto_wait=True,
+        logger=json_looger,  # here you can pass the custom logger to the trainer
```
Comment (tizianocitro) [inline examples/train/train_with_custom_logger.py]: Typo.

File: examples/train/train_with_early_stopping.py [hunk 3]
```diff
@@ -132,7 +132,7 @@
     ]
 
     early_stopping = EarlyStopping(
-        monitor="val/loss",
+        monitor="val_loss",
```
Comment (tizianocitro) [inline examples/train/train_with_early_stopping.py]: I believe this is "val/loss".

File: hyperbench/train/trainer.py [hunk 4]
```diff
@@ -263,7 +263,7 @@ def __trainer_for(
 
                 if should_create_test_trainer:
                     model_config.test_trainer = __trainer_for(
-                        trainer_devices=cast(list[int] | str | int, test_devices),
+                        trainer_devices=test_devices,
```
Comment (tizianocitro) [inline hyperbench/train/trainer.py]: We need to fix this in a PR first.

File: Makefile [hunk 5]
```diff
@@ -37,7 +37,7 @@ format:
 
 typecheck:
 	@echo '=== Type checking ==='
-	$(UV) run $(TYPECHECKER) check
+	$(UV) run $(TYPECHECKER) check --error-on-warning
```
Comment (tizianocitro) [inline Makefile:40]: Maybe this is good to do in the same PR where we fix `trainer.py`, so that this PR doesn't depend on this.

**Review by tizianocitro — CHANGES_REQUESTED** [review]

File: examples/splitting/custom_splitter.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    SamplingStrategy,
+)
+from hyperbench.types import HData
+
+
+class CustomSplitter(Splitter["Dataset", list["Dataset"]]):
+    def __init__(self):
+        super().__init__()
+
+    def split(self, to_split: Dataset, **kwargs) -> list[Dataset]:
```
Comment (tizianocitro) [inline examples/splitting/custom_splitter.py:14]: Something like this:

```python
    train_he = hdata.he[:, :num_train_incidences]
    val_he = hdata.he[:, :num_val_incidences]
    train_hdata = HData.from_hyperedge_index(train_he)
    val_hdata = HData.from_hyperedge_index(val_he)
```

**Review by tizianocitro — APPROVED** [review]
